### PR TITLE
implement :info

### PR DESCRIPTION
**Pull request by TBuc**
Should solve #30 

Performed simple tests:
```
 unknown / development version
         repl session

Type :help for a list of available commands.

> :info
Instruments: 
Current instruments: 
Events: 0
Markers: 0

> :load sample.alda
This action will overwrite the current score. Continue?
(y)es, (n)o
c> 
c> :info
Instruments: violin-xTKq3, violin-tmlVg, viola-dK98h, cello-vuFTY
Current instruments: cello-vuFTY
Events: 20
Markers: 0

c> :quit
```
where sample.alda contains:
```
violin "violin-1":
  o4 f2   g4 a   b-2   a

violin "violin-2":
  o4 c2   e4 f   f2    f

viola:
  o3 a2 > c4 c   d2    c

cello:
  (volume 75)
  o3 f2   c4 f < b-2 > f
```

Please test it for some edge-case inputs too (sorry but I have yet no idea what scores would do the job).
Any change request, just let me know.

### CODE REVIEW

File: src/alda/repl/commands/ReplInfo.java [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+  private Gson gson = new Gson();
+	
+  @Override
+  public void act(String args, StringBuffer history, AldaServer server,
+                  ConsoleReader reader, Consumer<AldaScore> newInstrument)
+                		  throws NoResponseException {
+	try {
+	  String res = server.parseRaw(history.toString(), "data");
+	  if(res == null) {
+	    System.err.println("An internal error occurred when reading the score.");
```
Comment (jgkamat) [inline src/alda/repl/commands/ReplInfo.java]: I think the indentation in this file is a bit wonky, this line has tabs and spaces.

File: src/alda/repl/commands/ReplInfo.java [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+    List<String> curInstrLst = gson.fromJson(curInstrJson, new TypeToken<List<String>>(){}.getType());
+	String instr = curInstrLst
+		.stream()
+		.collect(Collectors.joining(", "));
+	return instr;
+  }
+  
+  private int getMarkersAmount(JsonObject scoreMap) {
+    JsonObject markersJson = scoreMap.getAsJsonObject("markers");
+    int nMarkers = markersJson.entrySet().size() - 1;
```
Comment (jgkamat) [inline src/alda/repl/commands/ReplInfo.java]: Can you add a comment here, something along the lines of 'subtract by one to ignore the "start" marker'?

File: src/alda/repl/commands/ReplInfo.java [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+    		.append(System.lineSeparator());
+     
+        sb.append("Markers: ")
+    		.append(getMarkersAmount(scoreMap))
+    		.append(System.lineSeparator());
+        
+		System.out.println(sb);
+	  }
+	  
+	} catch(ParseError e) {
```
Comment (jgkamat) [inline src/alda/repl/commands/ReplInfo.java]: Since these clases are all the same catch block, I think you can:

```
catch(ParseError | JsonParseException | ClassCastException e) {
...
}
```

Why are you catching a classcastexception, by the way? I hope that isn't coming up here :P

**Review by jgkamat — COMMENTED** [review]
Overall, this looks awesome to me, and I think it would be a great addition! My comments are listed below :)

**Comment by daveyarwood** [comment]
I agree with @jgkamat's notes.

Testing this out locally, it looks great, but I realized that I think it would be better if, for "Markers:", instead of displaying the number of markers in the score, we listed the names in order by offset, ~~excluding start~~ _(EDIT: let's leave `start` in for now; I think I'd like to fix the backend so that it isn't including the `start` marker)_. For example:

```
Markers: verse, chorus, bridge
```

Another nice touch would be, if there are none of something (current instruments, instruments, markers), we display `(none)` as a placeholder.

File: src/alda/repl/commands/ReplInfo.java (hunk 1 above)
Comment (TBuc) [reply src/alda/repl/commands/ReplInfo.java]: You're right - my laziness :-P

File: src/alda/repl/commands/ReplInfo.java (hunk 2 above)
Comment (TBuc) [reply src/alda/repl/commands/ReplInfo.java]: Actually, as per Dave's request, the new version of this method won't need this anymore. BTW it was a good idea.

File: src/alda/repl/commands/ReplInfo.java (hunk 3 above)
Comment (TBuc) [reply src/alda/repl/commands/ReplInfo.java]: I've noticed that in the returned JSON "markers" and "instruments" are objects, while "current-instruments" and "events" are lists. Not sure why they are not all lists. This gave me a certain degree of uncertainty on the JSON structure contract reliability.
In case the contract changes in the future (i.e., one of such objects is changed to a list), a ClassCastException would be thrown.
But this was just due to some (wrong?) impression of mine, so if you say that the JSON structure is intended to be kept as is, I can remove it.

**Comment by TBuc** [comment]
Thank you both @jgkamat and @daveyarwood for your timely comments.
@daveyarwood great, I will modify accordingly to your request.

File: src/alda/repl/commands/ReplInfo.java (hunk 3 above)
Comment (daveyarwood) [reply src/alda/repl/commands/ReplInfo.java]: That's a good concern to have. I think we can reasonably rely on the shape of the data not changing, in this case. If we ever do want to change it for some reason, we'll just have to remember to adjust client code like this at the same time.

**Comment by TBuc** [comment]
Changes committed.
I've supposed marker's offset is an integer: I don't know Lisp, but the comment at line 9 in [offset.clj](https://github.com/alda-lang/alda-core/blob/a89fdde89a6fb3cfedef25173de2fdd887192770/src/alda/lisp/model/offset.clj) made me think of an integer value. Is this correct?

I could not find a score having multiple markers. In order to test the new markers feature, could you provide me with one, or tell me how to create markers in a score? I've looked up a bit with no luck.

File: src/alda/repl/commands/ReplInfo.java (hunk 3 above)
Comment (TBuc) [reply src/alda/repl/commands/ReplInfo.java]: Fixed.

File: src/alda/repl/commands/ReplInfo.java (hunk 1 above)
Comment (TBuc) [reply src/alda/repl/commands/ReplInfo.java]: Fixed.

**Comment by daveyarwood** [comment]
They are often integers, but can be floating-point values. Here is an example score:

```
piano:
  (tempo 93)
  c8 d e %one f %two g %three a
```

and the markers part of the score data:

```json
  "markers": {
    "start": 0,
    "one": 967.7419738769531,
    "two": 1290.3226318359375,
    "three": 1612.9032897949219
  },
```

**Comment by TBuc** [comment]
Changes tested with suggested input:
```
> :info
Instruments: (none)
Current instruments: (none)
Events: 0
Markers: start

> piano:
p>   (tempo 93)
p>   c8 d e %one f %two g %three a
p> :info
Instruments: piano-oV9Lu
Current instruments: piano-oV9Lu
Events: 6
Markers: start, one, two, three
```
I've tested it with a modified JSON input too:
```
"markers":{
  "start":0,
  "one":967.7419738769531,
  "three":11612.9032897949219,
  "two":1290.3226318359375
}
```
And seems to work fine:
```
p>:info
Instruments: piano-sFz6g
Current instruments: piano-sFz6g
Events: 6
Markers: start, one, two, three
```
In the next few days I will hopefully write some unit tests too. I've already splitted the act() method in order to enable passing arbitrary JSON input to the function for future test purposes.

**Comment by daveyarwood** [comment]
This is looking great so far! If you're up for writing unit tests*, that would be most welcome.

*or even integration tests -- we have a framework set up for that already in the `integrationtests` directory. That would be a good way for us to catch potential problems in the future if the server ever starts to send the data over with a different structure.

**Comment by daveyarwood** [comment]
NB: per discussion in Slack, we can merge this PR as-is and add tests in a later PR

Will merge this PR soon. Thanks again @TBuc! :+1:

**Review by daveyarwood — APPROVED** [review]
